File: codes/python/chapter_stack_and_queue/linkedlist_deque.py
"""
File: linkedlist_deque.py
Created Time: 2023-03-01
Author: Krahets (krahets@163.com)
"""

import os.path as osp
import sys

sys.path.append(osp.dirname(osp.dirname(osp.abspath(__file__))))
from include import *

""" 双向链表结点 """
class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None  # 后继结点引用（指针）
        self.prev = None  # 前驱结点引用（指针）

""" 基于双向链表实现的双向队列 """
class LinkedListDeque:
    """ 构造方法 """
    def __init__(self):
        self.front, self.rear = None, None  # 头结点 front ，尾结点 rear
        self.__size = 0  # 双向队列的长度

    """ 获取双向队列的长度 """
    def size(self):
        return self.__size

    """ 判断双向队列是否为空 """
    def is_empty(self):
        return self.size() == 0

    """ 入队操作 """
    def push(self, num, is_front):
        node = ListNode(num)
        # 若链表为空，则令 front, rear 都指向 node
        if self.is_empty():
            self.front = self.rear = node
        # 队首入队操作
        elif is_front:
            # 将 node 添加至链表头部
            self.front.prev = node
            node.next = self.front
            self.front = node  # 更新头结点
        # 队尾入队操作
        else:
            # 将 node 添加至链表尾部
            self.rear.next = node
            node.prev = self.rear
            self.rear = node  # 更新尾结点
        self.__size += 1  # 更新队列长度

    """ 队首入队 """
    def push_first(self, num):
        self.push(num, True)

    """ 队尾入队 """
    def push_last(self, num):
        self.push(num, False)

    """ 出队操作 """
    def poll(self, is_front):
        # 若队列为空，直接返回 None
        if self.is_empty():
            return None
        # 队首出队操作
        if is_front:
            val = self.front.val  # 暂存头结点值
            # 删除头结点
            fnext = self.front.next
            if fnext != None:
                fnext.prev = None
                self.front.next = None
            self.front = fnext  # 更新头结点
        # 队尾出队操作
        else:
            val = self.rear.val  # 暂存尾结点值
            # 删除尾结点
            rprev = self.rear.prev
            if rprev != None:
                rprev.next = None
                self.rear.prev = None
            self.rear = rprev  # 更新尾结点
        self.__size -= 1  # 更新队列长度
        return val

    """ 队首出队 """
    def poll_first(self):
        return self.poll(True)

    """ 队尾出队 """
    def poll_last(self):
        return self.poll(False)

    """ 访问队首元素 """
    def peek_first(self):
        return None if self.is_empty() else self.front.val

    """ 访问队尾元素 """
    def peek_last(self):
        return None if self.is_empty() else self.rear.val

    """ 返回数组用于打印 """
    def to_array(self):
        node = self.front
        res = [0] * self.size()
        for i in range(self.size()):
            res[i] = node.val
            node = node.next
        return res


""" Driver Code """
if __name__ == "__main__":
    """ 初始化双向队列 """
    deque = LinkedListDeque()
    deque.push_last(3)
    deque.push_last(2)
    deque.push_last(5)
    print("双向队列 deque =", deque.to_array())

    """ 访问元素 """
    peek_first = deque.peek_first()
    print("队首元素 peek_first =", peek_first)
    peek_last = deque.peek_last()
    print("队尾元素 peek_last =", peek_last)

    """ 元素入队 """
    deque.push_last(4)
    print("元素 4 队尾入队后 deque =", deque.to_array())
    deque.push_first(1)
    print("元素 1 队首入队后 deque =", deque.to_array())

    """ 元素出队 """
    poll_last = deque.poll_last()
    print("队尾出队元素 =", poll_last, "，队尾出队后 deque =", deque.to_array())
    poll_first = deque.poll_first()
    print("队首出队元素 =", poll_first, "，队首出队后 deque =", deque.to_array())

    """ 获取双向队列的长度 """
    size = deque.size()
    print("双向队列长度 size =", size)

    """ 判断双向队列是否为空 """
    is_empty = deque.is_empty()
    print("双向队列是否为空 =", is_empty)

File: codes/javascript/chapter_stack_and_queue/linkedlist_deque.js
/**
 * File: linkedlist_deque.js
 * Created Time: 2023-02-04
 * Author: Zhuo Qinyue (1403450829@qq.com)
 */

/* 双向链表结点 */
class ListNode {
    prev;   // 前驱结点引用 (指针)
    next;   // 后继结点引用 (指针)
    val;    // 结点值
    
    constructor(val) {
        this.val = val;
        this.next = null;
        this.prev = null;
    }
}

/* 基于双向链表实现的双向队列 */
class LinkedListDeque {
    front;  // 头结点 front
    rear;   // 尾结点 rear
    len;    // 双向队列的长度

    constructor() {
        this.front = null;
        this.rear = null;
        this.len = 0;
    }

    /* 队尾入队操作 */
    pushLast(val) {
        const node = new ListNode(val);
        // 若链表为空，则令 front, rear 都指向 node
        if (this.len === 0) {
            this.front = node;
            this.rear = node;
        } else {
            // 将 node 添加至链表尾部
            this.rear.next = node;
            node.prev = this.rear;
            this.rear = node; // 更新尾结点
        }
        this.len++;
    }

    /* 队首入队操作 */
    pushFirst(val) {
        const node = new ListNode(val);
        // 若链表为空，则令 front, rear 都指向 node
        if (this.len === 0) {
            this.front = node;
            this.rear = node;
        } else {
            // 将 node 添加至链表头部
            this.front.prev = node;
            node.next = this.front;
            this.front = node; // 更新头结点
        }
        this.len++;
    }

    /* 队尾出队操作 */
    pollLast() {
        if (this.len === 0) {
            return null;
        }
        const value = this.rear.val; // 存储尾结点值
        // 删除尾结点
        let temp = this.rear.prev;
        if (temp !== null) {
            temp.next = null;
            this.rear.prev = null;
        }
        this.rear = temp;   // 更新尾结点
        this.len--;
        return value;
    }

    /* 队首出队操作 */
    pollFirst() {
        if (this.len === 0) {
            return null;
        }
        const value = this.front.val; // 存储尾结点值
        // 删除头结点
        let temp = this.front.next;
        if (temp !== null) {
            temp.prev = null;
            this.front.next = null;
        }
        this.front = temp;   // 更新头结点
        this.len--;
        return value;
    }

    /* 访问队尾元素 */
    peekLast() {
        return this.len === 0 ? null : this.rear.val;
    }

    /* 访问队首元素 */
    peekFirst() {
        return this.len === 0 ? null : this.front.val;
    }

    /* 获取双向队列的长度 */
    size() {
        return this.len;
    }

    /* 判断双向队列是否为空 */
    isEmpty() {
        return this.len === 0;
    }

    /* 打印双向队列 */
    print() {
        const arr = [];
        let temp = this.front;
        while (temp !== null) {
            arr.push(temp.val);
            temp = temp.next;
        }
        console.log("[" + arr.join(", ") + "]");
    }
}

/* 初始化双向队列 */
const linkedListDeque = new LinkedListDeque();
linkedListDeque.pushLast(3);
linkedListDeque.pushLast(2);
linkedListDeque.pushLast(5);
console.log("双向队列 linkedListDeque = ");
linkedListDeque.print();

/* 访问元素 */
const peekFirst = linkedListDeque.peekFirst();
console.log("队首元素 peekFirst = " + peekFirst);
const peekLast = linkedListDeque.peekLast();
console.log("队尾元素 peekLast = " + peekLast);

/* 元素入队 */
linkedListDeque.pushLast(4);
console.log("元素 4 队尾入队后 linkedListDeque = ");
linkedListDeque.print();
linkedListDeque.pushFirst(1);
console.log("元素 1 队首入队后 linkedListDeque = ");
linkedListDeque.print();

/* 元素出队 */
const pollLast = linkedListDeque.pollLast();
console.log("队尾出队元素 = " + pollLast + "，队尾出队后 linkedListDeque = ");
linkedListDeque.print();
const pollFirst = linkedListDeque.pollFirst();
console.log("队首出队元素 = " + pollFirst + "，队首出队后 linkedListDeque = ");
linkedListDeque.print();

/* 获取双向队列的长度 */
const size = linkedListDeque.size();
console.log("双向队列长度 size = " + size);

/* 判断双向队列是否为空 */
const isEmpty = linkedListDeque.isEmpty();
console.log("双向队列是否为空 = " + isEmpty);
